source/distributed/fullydistributed_tria.cc

```diff
@@ -0,0 +1,562 @@
+            {
+              case 1:
+                vertices[0] = p1;
+                vertices[1] = p2;
+                break;
+              case 2:
+                vertices[0] = vertices[1] = p1;
+                vertices[2] = vertices[3] = p2;
+
+                vertices[1](0) = p2(0);
+                vertices[2](0) = p1(0);
+                break;
+              case 3:
+                vertices[0] = vertices[1] = vertices[2] = vertices[3] = p1;
+                vertices[4] = vertices[5] = vertices[6] = vertices[7] = p2;
+
+                vertices[1](0) = p2(0);
+                vertices[2](1) = p2(1);
+                vertices[3](0) = p2(0);
+                vertices[3](1) = p2(1);
+
+                vertices[4](0) = p1(0);
+                vertices[4](1) = p1(1);
+                vertices[5](1) = p1(1);
+                vertices[6](0) = p1(0);
+
+
+                break;
+              default:
+                Assert(false, ExcNotImplemented());
+            }
+
+          std::vector<CellData<dim>> cells(1);
+          for (unsigned int i = 0; i < GeometryInfo<dim>::vertices_per_cell;
+               ++i)
+            cells[0].vertices[i] = i;
+          cells[0].material_id = 0;
+          SubCellData subcelldata;
+          dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+            vertices, cells, subcelldata);
+
+          // 2) mark cell as artificial
+          for (auto cell = this->begin(); cell != this->end(); cell++)
+            {
+              cell->set_subdomain_id(dealii::numbers::artificial_subdomain_id);
+              cell->set_level_subdomain_id(
+                dealii::numbers::artificial_subdomain_id);
+            }
+
+          // 3) setup dummy mapping between locally relevant coarse-grid cells
+          //    and global cells
+          this->coarse_gid_to_lid.emplace_back(0, 0);
+          this->coarse_lid_to_gid.emplace_back(0, 0);
+        }
+      else
+        {
+          auto &boundary_ids = construction_data.boundary_ids;
+          auto &cells        = construction_data.cells;
+          auto &vertices     = construction_data.vertices;
+          auto &parts        = construction_data.parts;
+
+
+
+          // create inverse map
+          std::map<int, int> coarse_gid_to_lid;
+          for (auto &i : coarse_lid_to_gid)
+            coarse_gid_to_lid[i.second] = i.first;
+
+
+
+          // convert map to vector and save data structures
+          for (auto i : construction_data.coarse_lid_to_gid)
+            this->coarse_lid_to_gid.emplace_back(i);
+
+          for (auto i : coarse_gid_to_lid)
+            this->coarse_gid_to_lid.emplace_back(i);
+
+
+
+          // 1) create coarse grid
+          SubCellData subcelldata;
+          dealii::parallel::Triangulation<dim, spacedim>::create_triangulation(
+            vertices, cells, subcelldata);
+
+          for (auto c = this->begin(); c != this->end(); c++)
+            c->set_all_manifold_ids(c->material_id());
```

Comment:
why do you do this? Can you make sure this is documented in the class?

Reply:
Ups. What I meant (I guess my test passed because `material_id` and `manifold_id` happened to be the same):
```cpp
          for (auto c = this->begin(); c != this->end(); c++)
            c->set_all_manifold_ids(c->manifold_id());
```

I need theses two lines to be able to create a `hyper_shell()` with vertices put at the right position while refining (see also dealii step-1). Setting `CellData::manifold_id` was not enough...

Reply:
The user is free to use different manifolds ids on cells, faces, and edges (and this is often useful to do!). The call to ``set_all_manifold_ids`` will set this for the cell and all child objects. I assume your ConstructionData does not carry information about face/edge manifold ids? I think it should...

Reply:
OK! I will extend `ConstructionData` with this information.

Reply:
@tjhei I have extended `ConstructionData` with `manifold_quad_ids` and `manifold_line_ids` (and have added some comments). 

Could you please have a look if this is fine this way?